File: sim/outfield_decision.lua
-- Serializable personal decision cadence and carrier-option construction.
-- Callers own world geometry and execution; this module retains only value
-- state and resolves deterministic numeric contexts through sim.brain.

local brain = require("sim.brain")
local rng = require("core.rng")

---@alias OutfieldDecisionContext "ineligible"|"offball"|"carrier"
---@alias OutfieldIntent "none"|"move"|"in_behind"|"come_short"|"hold_width"|"shoot"|"cross"|"pass"|"dribble"

---@class OutfieldDecisionState
---@field version integer
---@field generation integer
---@field rng_state integer
---@field remaining number
---@field context OutfieldDecisionContext
---@field intent OutfieldIntent
---@field target_x number?
---@field target_y number?
---@field target_player integer?
---@field run_expires_at number?

---@class OutfieldPassOption
---@field player_index integer
---@field openness number
---@field forward_progress number
---@field distance number
---@field lane_blocked boolean
---@field interception_risk boolean
---@field lane_fraction number?

---@class OutfieldCarrierContext
---@field goal_distance number
---@field shoot_range number
---@field angle_quality number
---@field keeper_coverage number
---@field space number
---@field flank_depth number
---@field cross_target integer?
---@field box_targets integer
---@field cross_space number
---@field goal_progress number
---@field dribble_space number
---@field passes OutfieldPassOption[]

---@class OutfieldDecisionModule
local outfield_decision = {}

outfield_decision.VERSION = 2
outfield_decision.SLOW_REFRESH_SECONDS = 0.45
outfield_decision.FAST_REFRESH_SECONDS = 0.15
outfield_decision.BASE_TEMPERATURE = 3
outfield_decision.RUN_LIFETIME_SECONDS = 1.8

local SHARP_COMPOSURE = 0.8
local PASS_RISK_PENALTY = 5
local PASS_SPACE_BALANCE = 0.6
local PASS_SPACE_WEIGHT = 90

local CONTEXTS = {
    ineligible = true,
    offball = true,
    carrier = true,
}

local INTENTS = {
    none = true,
    move = true,
    in_behind = true,
    come_short = true,
    hold_width = true,
    shoot = true,
    cross = true,
    pass = true,
    dribble = true,
}

local RUN_INTENTS = {
    in_behind = true,
    come_short = true,
    hold_width = true,
}

local CARRIER_INTENTS = {
    shoot = true,
    cross = true,
    pass = true,
    dribble = true,
}

---@param value number
---@return boolean
local function is_finite(value)
    return value == value and value ~= math.huge and value ~= -math.huge
end

---@param value number
---@param label string
local function assert_finite(value, label)
    assert(type(value) == "number" and is_finite(value), label .. " must be finite")
end

---@param value number
---@return number
local function clamp_unit(value)
    if value ~= value or value == -math.huge then
        return 0
    end
    if value == math.huge then
        return 1
    end
    return math.max(0, math.min(1, value))
end

---@param rng_state integer?
---@return OutfieldDecisionState
function outfield_decision.new_state(rng_state)
    return {
        version = outfield_decision.VERSION,
        generation = 0,
        rng_state = rng.seed(rng_state or 1),
        remaining = 0,
        context = "ineligible",
        intent = "none",
        target_x = nil,
        target_y = nil,
        target_player = nil,
        run_expires_at = nil,
    }
end

---@param state OutfieldDecisionState
local function validate_state(state)
    assert(type(state) == "table", "outfield decision state must be a table")
    assert(state.version == outfield_decision.VERSION, "unsupported outfield decision version")
    assert(
        type(state.generation) == "number"
            and is_finite(state.generation)
            and state.generation == math.floor(state.generation)
            and state.generation >= 0,
        "outfield decision generation must be a non-negative integer"
    )
    assert(
        type(state.rng_state) == "number"
            and is_finite(state.rng_state)
            and state.rng_state == math.floor(state.rng_state)
            and rng.seed(state.rng_state) == state.rng_state,
        "outfield decision RNG state must be a canonical positive integer"
    )
    assert_finite(state.remaining, "outfield decision remaining time")
    assert(state.remaining >= 0, "outfield decision remaining time must be non-negative")
    assert(CONTEXTS[state.context], "unknown outfield decision context")
    assert(INTENTS[state.intent], "unknown outfield decision intent")
    if state.target_x ~= nil then
        assert_finite(state.target_x, "outfield decision target x")
    end
    if state.target_y ~= nil then
        assert_finite(state.target_y, "outfield decision target y")
    end
    assert(
        (state.target_x == nil) == (state.target_y == nil),
        "outfield decision target coordinates must be paired"
    )
    if state.target_player ~= nil then
        assert(
            type(state.target_player) == "number"
                and is_finite(state.target_player)
                and state.target_player == math.floor(state.target_player)
                and state.target_player >= 1,
            "outfield decision target player must be a positive integer"
        )
    end
    if state.run_expires_at ~= nil then
        assert_finite(state.run_expires_at, "outfield decision run expiry")
    end
    if state.context == "ineligible" then
        assert(state.intent == "none", "ineligible outfield decision state must have no intent")
        assert(state.remaining == 0, "ineligible outfield decision state cannot retain cadence")
        assert(
            state.target_x == nil and state.target_player == nil,
            "ineligible outfield decision state cannot retain a target"
        )
        assert(
            state.run_expires_at == nil,
            "ineligible outfield decision state cannot retain a run expiry"
        )
    elseif state.context == "offball" then
        assert(
            state.intent == "move" or RUN_INTENTS[state.intent],
            "offball outfield decision state must retain a movement intent"
        )
        assert(state.target_x ~= nil, "offball movement intent requires a target point")
        assert(state.target_player == nil, "offball movement intent cannot target a player")
        if RUN_INTENTS[state.intent] then
            assert(state.run_expires_at ~= nil, "offball run intent requires an absolute expiry")
        else
            assert(
                state.run_expires_at == nil,
                "ordinary offball movement cannot retain a run expiry"
            )
        end
    else
        assert(state.context == "carrier", "unknown eligible outfield decision context")
        assert(
            CARRIER_INTENTS[state.intent],
            "carrier outfield decision state requires a carrier intent"
        )
        if state.intent == "pass" or state.intent == "cross" then
            assert(state.target_player ~= nil, "pass and cross intents require a target player")
            assert(state.target_x == nil, "pass and cross intents cannot retain a target point")
        else
            assert(
                state.target_x == nil and state.target_player == nil,
                "shoot and dribble intents cannot retain a target"
            )
        end
        assert(
            state.run_expires_at == nil,
            "carrier outfield decision state cannot retain a run expiry"
        )
    end
end

---@param intent OutfieldIntent
---@return boolean
function outfield_decision.is_run_intent(intent)
    return RUN_INTENTS[intent] == true
end

---@param state OutfieldDecisionState
---@return OutfieldDecisionState
function outfield_decision.copy_state(state)
    validate_state(state)
    return {
        version = state.version,
        generation = state.generation,
        rng_state = state.rng_state,
        remaining = state.remaining,
        context = state.context,
        intent = state.intent,
        target_x = state.target_x,
        target_y = state.target_y,
        target_player = state.target_player,
        run_expires_at = state.run_expires_at,
    }
end

---@param state OutfieldDecisionState
---@param dt number
---@return OutfieldDecisionState
function outfield_decision.advance(state, dt)
    local next_state = outfield_decision.copy_state(state)
    assert_finite(dt, "outfield decision delta")
    assert(dt >= 0, "outfield decision delta must be non-negative")
    next_state.remaining = math.max(0, next_state.remaining - dt)
    return next_state
end

---@param state OutfieldDecisionState
---@return OutfieldDecisionState
function outfield_decision.reset(state)
    local current = outfield_decision.copy_state(state)
    local reset = outfield_decision.new_state(current.rng_state)
    reset.generation = current.generation
    return reset
end

---@param state OutfieldDecisionState
---@param rng_state integer
---@return OutfieldDecisionState
function outfield_decision.with_rng_state(state, rng_state)
    local next_state = outfield_decision.copy_state(state)
    next_state.rng_state = rng_state
    return outfield_decision.copy_state(next_state)
end

-- End an active run without inventing a cadence boundary. The caller supplies
-- the ordinary support fallback that takes over for the retained countdown.
---@param state OutfieldDecisionState
---@param target_x number
---@param target_y number
---@return OutfieldDecisionState
function outfield_decision.cancel_run(state, target_x, target_y)
    local next_state = outfield_decision.copy_state(state)
    assert(RUN_INTENTS[next_state.intent], "cancel_run requires an active run intent")
    next_state.intent = "move"
    next_state.target_x = target_x
    next_state.target_y = target_y
    next_state.target_player = nil
    next_state.run_expires_at = nil
    return outfield_decision.copy_state(next_state)
end

---@param state OutfieldDecisionState
---@param context OutfieldDecisionContext
---@param urgent boolean?
---@return boolean
function outfield_decision.should_refresh(state, context, urgent)
    validate_state(state)
    assert(CONTEXTS[context], "unknown requested outfield decision context")
    return context ~= "ineligible"
        and (urgent == true or state.context ~= context or state.remaining <= 0)
end

---@param state OutfieldDecisionState
---@param context OutfieldDecisionContext
---@param intent OutfieldIntent
---@param scan_rate number
---@param target_x number?
---@param target_y number?
---@param target_player integer?
---@param run_expires_at number?
---@return OutfieldDecisionState
function outfield_decision.refresh(
    state,
    context,
    intent,
    scan_rate,
    target_x,
    target_y,
    target_player,
    run_expires_at
)
    validate_state(state)
    assert(context ~= "ineligible" and CONTEXTS[context], "refresh requires an eligible context")
    local refreshed = {
        version = outfield_decision.VERSION,
        generation = state.generation + 1,
        rng_state = state.rng_state,
        remaining = brain.refresh_interval(
            scan_rate,
            outfield_decision.SLOW_REFRESH_SECONDS,
            outfield_decision.FAST_REFRESH_SECONDS
        ),
        context = context,
        intent = intent,
        target_x = target_x,
        target_y = target_y,
        target_player = target_player,
        run_expires_at = run_expires_at,
    }
    return outfield_decision.copy_state(refreshed)
end

---@param context OutfieldCarrierContext
---@return CarrierOption[]
function outfield_decision.carrier_options(context)
    assert_finite(context.goal_distance, "carrier goal distance")
    assert(context.goal_distance >= 0, "carrier goal distance must be non-negative")
    assert_finite(context.shoot_range, "carrier shoot range")
    assert(context.shoot_range > 0, "carrier shoot range must be positive")
    assert_finite(context.space, "carrier space")
    assert_finite(context.cross_space, "carrier cross space")
    assert_finite(context.flank_depth, "carrier flank depth")
    assert_finite(context.goal_progress, "carrier goal progress")
    assert_finite(context.dribble_space, "carrier dribble space")
    assert(
        type(context.box_targets) == "number"
            and context.box_targets == math.floor(context.box_targets)
            and context.box_targets >= 0,
        "carrier box target count must be a non-negative integer"
    )

    local distance_falloff = context.goal_distance / context.shoot_range
    local options = {
        {
            id = "shoot",
            kind = "shoot",
            score = 100 - distance_falloff * distance_falloff * 60 + clamp_unit(
                context.angle_quality
            ) * 25 - clamp_unit(context.keeper_coverage) * 12 + clamp_unit(
                context.space
            ) * 18,
        },
        {
            id = "dribble",
            kind = "dribble",
            score = 10
                + clamp_unit(context.dribble_space) * 35
                + clamp_unit(context.goal_progress) * 25,
        },
    }

    if context.cross_target ~= nil and context.box_targets > 0 then
        options[#options + 1] = {
            id = "cross",
            kind = "cross",
            score = 60
                + clamp_unit(context.flank_depth) * 45
                + math.min(3, context.box_targets) * 8
                + clamp_unit(context.cross_space) * 15,
            reference = context.cross_target,
        }
    end

    local seen = {}
    for index, pass in ipairs(context.passes) do
        assert(
            type(pass.player_index) == "number"
                and pass.player_index == math.floor(pass.player_index)
                and pass.player_index >= 1,
            "pass option " .. index .. " player must be a positive integer"
        )
        assert(not seen[pass.player_index], "pass option players must be unique")
        seen[pass.player_index] = true
        assert_finite(pass.openness, "pass option openness")
        assert_finite(pass.forward_progress, "pass option forward progress")
        assert_finite(pass.distance, "pass option distance")
        assert(pass.distance >= 0, "pass option distance must be non-negative")
        if pass.lane_fraction ~= nil then
            assert_finite(pass.lane_fraction, "pass option lane fraction")
        end
        options[#options + 1] = {
            id = "pass_" .. tostring(pass.player_index),
            kind = "pass",
            score = 18 + math.min(1, math.max(0, pass.openness) / 150) * 45 + math.max(
                -1,
                math.min(1, pass.forward_progress / 300)
            ) * 28 - math.min(1, pass.distance / 420) * 18 + (PASS_SPACE_BALANCE - clamp_unit(
                context.space
            )) * PASS_SPACE_WEIGHT - (pass.interception_risk and PASS_RISK_PENALTY or 0),
            payload = pass.lane_fraction and { lane_fraction = pass.lane_fraction } or nil,
            reference = pass.player_index,
        }
    end
    return options
end

---@param options CarrierOption[]
---@param composure number
---@param pressure number
---@param rng_state integer
---@return CarrierOption selected
---@return integer next_rng_state
function outfield_decision.decide_carrier(options, composure, pressure, rng_state)
    -- At the authored high-composure boundary the temperature is exactly zero:
    -- an effectively settled choice takes the argmax without needlessly
    -- advancing its dedicated decision RNG stream.
    local base_temperature = composure >= SHARP_COMPOSURE and 0
        or outfield_decision.BASE_TEMPERATURE
    return brain.decide_carrier(options, composure, pressure, base_temperature, rng_state)
end

return outfield_decision

File: rust/crates/gc-sim/src/outfield_decision.rs
//! Port of `sim/outfield_decision.lua`.
//!
//! Serializable personal decision cadence and carrier-option construction.
//! Callers own world geometry and execution; this module retains only value
//! state and resolves deterministic numeric contexts through
//! [`crate::brain`].

use crate::brain;
use gc_core::rng;
use indexmap::IndexMap;

/// Which eligibility context a decision was made under.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum OutfieldDecisionContext {
    /// The player is not eligible to act right now.
    Ineligible,
    /// The player is off the ball.
    Offball,
    /// The player is carrying the ball.
    Carrier,
}

/// What a player has decided to do.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum OutfieldIntent {
    /// No decision has been made yet.
    None,
    /// Move to an ordinary support point.
    Move,
    /// Run in behind the last defender.
    InBehind,
    /// Come short to offer an outlet.
    ComeShort,
    /// Hold width out on the flank.
    HoldWidth,
    /// Shoot at goal.
    Shoot,
    /// Cross into the box.
    Cross,
    /// Pass to a teammate.
    Pass,
    /// Dribble with the ball.
    Dribble,
}

/// This module's serializable per-player decision state.
#[derive(Clone, Copy, Debug, PartialEq)]
pub struct OutfieldDecisionState {
    /// This state's format version.
    pub version: u32,
    /// Monotonic count of every [`refresh`] this state has been through.
    pub generation: u32,
    /// This player's dedicated decision RNG stream state.
    pub rng_state: u32,
    /// Seconds remaining before the next refresh is due.
    pub remaining: f64,
    /// The eligibility context this decision was made under.
    pub context: OutfieldDecisionContext,
    /// The decided intent.
    pub intent: OutfieldIntent,
    /// The decision's target x, if it targets a point.
    pub target_x: Option<f64>,
    /// The decision's target y, if it targets a point.
    pub target_y: Option<f64>,
    /// The decision's target player, if it targets a player.
    pub target_player: Option<u32>,
    /// The absolute expiry of an active run, if this is a run intent.
    pub run_expires_at: Option<f64>,
}

/// One pass option offered to [`carrier_options`].
#[derive(Clone, Copy, Debug, PartialEq)]
pub struct OutfieldPassOption {
    /// The receiving player's stable identity.
    pub player_index: u32,
    /// How open the receiver is.
    pub openness: f64,
    /// How much forward progress the pass makes.
    pub forward_progress: f64,
    /// The pass distance.
    pub distance: f64,
    /// Whether the caller has determined the lane is blocked.
    pub lane_blocked: bool,
    /// Whether the caller has flagged this pass as interceptable.
    pub interception_risk: bool,
    /// Where along the lane a blocker sits, if any.
    pub lane_fraction: Option<f64>,
}

/// Inputs to [`carrier_options`].
#[derive(Clone, Debug, PartialEq)]
pub struct OutfieldCarrierContext {
    /// Distance to the goal.
    pub goal_distance: f64,
    /// The carrier's effective shooting range. Must be positive.
    pub shoot_range: f64,
    /// How good the shooting angle is, in `[0, 1]`.
    pub angle_quality: f64,
    /// How well the keeper covers the goal, in `[0, 1]`.
    pub keeper_coverage: f64,
    /// How much space the carrier has, in `[0, 1]`.
    pub space: f64,
    /// How deep the flank runner is, in `[0, 1]`.
    pub flank_depth: f64,
    /// The best cross target, if any.
    pub cross_target: Option<u32>,
    /// How many teammates are in the box.
    pub box_targets: u32,
    /// How much space the cross target has, in `[0, 1]`.
    pub cross_space: f64,
    /// How much goal progress a dribble would make, in `[0, 1]`.
    pub goal_progress: f64,
    /// How much space a dribble has, in `[0, 1]`.
    pub dribble_space: f64,
    /// Every pass option available to the carrier.
    pub passes: Vec<OutfieldPassOption>,
}

/// This module's format version.
pub const VERSION: u32 = 2;
/// The refresh interval at zero scan rate.
pub const SLOW_REFRESH_SECONDS: f64 = 0.45;
/// The refresh interval at full scan rate.
pub const FAST_REFRESH_SECONDS: f64 = 0.15;
/// The base softmax temperature for an ordinary carrier decision.
pub const BASE_TEMPERATURE: f64 = 3.0;
/// How long a granted run stays active, in seconds.
pub const RUN_LIFETIME_SECONDS: f64 = 1.8;

const SHARP_COMPOSURE: f64 = 0.8;
const PASS_RISK_PENALTY: f64 = 5.0;
const PASS_SPACE_BALANCE: f64 = 0.6;
const PASS_SPACE_WEIGHT: f64 = 90.0;

fn clamp_unit(value: f64) -> f64 {
    if value.is_nan() || value == f64::NEG_INFINITY {
        return 0.0;
    }
    if value == f64::INFINITY {
        return 1.0;
    }
    value.clamp(0.0, 1.0)
}

/// Whether `intent` is one of the off-ball run intents.
#[must_use]
pub fn is_run_intent(intent: OutfieldIntent) -> bool {
    matches!(
        intent,
        OutfieldIntent::InBehind | OutfieldIntent::ComeShort | OutfieldIntent::HoldWidth
    )
}

fn is_carrier_intent(intent: OutfieldIntent) -> bool {
    matches!(
        intent,
        OutfieldIntent::Shoot
            | OutfieldIntent::Cross
            | OutfieldIntent::Pass
            | OutfieldIntent::Dribble
    )
}

fn validate_state(state: &OutfieldDecisionState) {
    assert_eq!(
        state.version, VERSION,
        "unsupported outfield decision version"
    );
    assert_eq!(
        rng::seed(f64::from(state.rng_state)),
        state.rng_state,
        "outfield decision RNG state must be a canonical positive integer"
    );
    assert!(
        state.remaining.is_finite(),
        "outfield decision remaining time must be finite"
    );
    assert!(
        state.remaining >= 0.0,
        "outfield decision remaining time must be non-negative"
    );
    if let Some(x) = state.target_x {
        assert!(x.is_finite(), "outfield decision target x must be finite");
    }
    if let Some(y) = state.target_y {
        assert!(y.is_finite(), "outfield decision target y must be finite");
    }
    assert_eq!(
        state.target_x.is_none(),
        state.target_y.is_none(),
        "outfield decision target coordinates must be paired"
    );
    if let Some(target_player) = state.target_player {
        assert!(
            target_player >= 1,
            "outfield decision target player must be a positive integer"
        );
    }
    if let Some(run_expires_at) = state.run_expires_at {
        assert!(
            run_expires_at.is_finite(),
            "outfield decision run expiry must be finite"
        );
    }
    match state.context {
        OutfieldDecisionContext::Ineligible => {
            assert_eq!(
                state.intent,
                OutfieldIntent::None,
                "ineligible outfield decision state must have no intent"
            );
            assert_eq!(
                state.remaining, 0.0,
                "ineligible outfield decision state cannot retain cadence"
            );
            assert!(
                state.target_x.is_none() && state.target_player.is_none(),
                "ineligible outfield decision state cannot retain a target"
            );
            assert!(
                state.run_expires_at.is_none(),
                "ineligible outfield decision state cannot retain a run expiry"
            );
        }
        OutfieldDecisionContext::Offball => {
            assert!(
                state.intent == OutfieldIntent::Move || is_run_intent(state.intent),
                "offball outfield decision state must retain a movement intent"
            );
            assert!(
                state.target_x.is_some(),
                "offball movement intent requires a target point"
            );
            assert!(
                state.target_player.is_none(),
                "offball movement intent cannot target a player"
            );
            if is_run_intent(state.intent) {
                assert!(
                    state.run_expires_at.is_some(),
                    "offball run intent requires an absolute expiry"
                );
            } else {
                assert!(
                    state.run_expires_at.is_none(),
                    "ordinary offball movement cannot retain a run expiry"
                );
            }
        }
        OutfieldDecisionContext::Carrier => {
            assert!(
                is_carrier_intent(state.intent),
                "carrier outfield decision state requires a carrier intent"
            );
            if state.intent == OutfieldIntent::Pass || state.intent == OutfieldIntent::Cross {
                assert!(
                    state.target_player.is_some(),
                    "pass and cross intents require a target player"
                );
                assert!(
                    state.target_x.is_none(),
                    "pass and cross intents cannot retain a target point"
                );
            } else {
                assert!(
                    state.target_x.is_none() && state.target_player.is_none(),
                    "shoot and dribble intents cannot retain a target"
                );
            }
            assert!(
                state.run_expires_at.is_none(),
                "carrier outfield decision state cannot retain a run expiry"
            );
        }
    }
}

/// A fresh, ineligible decision state, seeded from a raw (not necessarily
/// canonical) RNG seed. Defaults to seed `1.0` when `rng_seed` is `None`.
#[must_use]
pub fn new_state(rng_seed: Option<f64>) -> OutfieldDecisionState {
    OutfieldDecisionState {
        version: VERSION,
        generation: 0,
        rng_state: rng::seed(rng_seed.unwrap_or(1.0)),
        remaining: 0.0,
        context: OutfieldDecisionContext::Ineligible,
        intent: OutfieldIntent::None,
        target_x: None,
        target_y: None,
        target_player: None,
        run_expires_at: None,
    }
}

/// Validate and defensively copy a decision state.
#[must_use]
pub fn copy_state(state: &OutfieldDecisionState) -> OutfieldDecisionState {
    validate_state(state);
    *state
}

/// Advance the cadence countdown by `dt` seconds, floored at zero.
#[must_use]
pub fn advance(state: &OutfieldDecisionState, dt: f64) -> OutfieldDecisionState {
    let mut next_state = copy_state(state);
    assert!(dt.is_finite(), "outfield decision delta must be finite");
    assert!(dt >= 0.0, "outfield decision delta must be non-negative");
    next_state.remaining = (next_state.remaining - dt).max(0.0);
    next_state
}

/// Drop back to the ineligible baseline, keeping the RNG stream and the
/// refresh generation counter.
#[must_use]
pub fn reset(state: &OutfieldDecisionState) -> OutfieldDecisionState {
    let current = copy_state(state);
    let mut reset_state = new_state(Some(f64::from(current.rng_state)));
    reset_state.generation = current.generation;
    reset_state
}

/// Replace the RNG stream state (e.g. after a caller advances it via
/// [`decide_carrier`]).
#[must_use]
pub fn with_rng_state(state: &OutfieldDecisionState, rng_state: u32) -> OutfieldDecisionState {
    let mut next_state = copy_state(state);
    next_state.rng_state = rng_state;
    copy_state(&next_state)
}

/// End an active run without inventing a cadence boundary. The caller
/// supplies the ordinary support fallback that takes over for the retained
/// countdown.
///
/// # Panics
///
/// Panics if `state` does not hold an active run intent.
#[must_use]
pub fn cancel_run(
    state: &OutfieldDecisionState,
    target_x: f64,
    target_y: f64,
) -> OutfieldDecisionState {
    let mut next_state = copy_state(state);
    assert!(
        is_run_intent(next_state.intent),
        "cancel_run requires an active run intent"
    );
    next_state.intent = OutfieldIntent::Move;
    next_state.target_x = Some(target_x);
    next_state.target_y = Some(target_y);
    next_state.target_player = None;
    next_state.run_expires_at = None;
    copy_state(&next_state)
}

/// Whether a fresh decision is due: an ineligible context never refreshes;
/// otherwise a context change, an urgent request, or an expired cadence
/// forces one.
#[must_use]
pub fn should_refresh(
    state: &OutfieldDecisionState,
    context: OutfieldDecisionContext,
    urgent: Option<bool>,
) -> bool {
    validate_state(state);
    context != OutfieldDecisionContext::Ineligible
        && (urgent == Some(true) || state.context != context || state.remaining <= 0.0)
}

/// Commit a fresh decision: bumps the generation, resets the cadence from
/// `scan_rate`, and records the new context/intent/target.
///
/// # Panics
///
/// Panics if `context` is [`OutfieldDecisionContext::Ineligible`].
#[must_use]
#[allow(clippy::too_many_arguments)]
pub fn refresh(
    state: &OutfieldDecisionState,
    context: OutfieldDecisionContext,
    intent: OutfieldIntent,
    scan_rate: f64,
    target_x: Option<f64>,
    target_y: Option<f64>,
    target_player: Option<u32>,
    run_expires_at: Option<f64>,
) -> OutfieldDecisionState {
    validate_state(state);
    assert!(
        context != OutfieldDecisionContext::Ineligible,
        "refresh requires an eligible context"
    );
    let refreshed = OutfieldDecisionState {
        version: VERSION,
        generation: state.generation + 1,
        rng_state: state.rng_state,
        remaining: brain::refresh_interval(scan_rate, SLOW_REFRESH_SECONDS, FAST_REFRESH_SECONDS),
        context,
        intent,
        target_x,
        target_y,
        target_player,
        run_expires_at,
    };
    copy_state(&refreshed)
}

/// Build the carrier's complete legitimate option set: shoot, dribble, an
/// optional cross, then one pass option per candidate. Construction consumes
/// no RNG; only [`decide_carrier`] does.
#[must_use]
pub fn carrier_options(context: &OutfieldCarrierContext) -> Vec<brain::CarrierOption> {
    assert!(
        context.goal_distance.is_finite(),
        "carrier goal distance must be finite"
    );
    assert!(
        context.goal_distance >= 0.0,
        "carrier goal distance must be non-negative"
    );
    assert!(
        context.shoot_range.is_finite(),
        "carrier shoot range must be finite"
    );
    assert!(
        context.shoot_range > 0.0,
        "carrier shoot range must be positive"
    );
    assert!(context.space.is_finite(), "carrier space must be finite");
    assert!(
        context.cross_space.is_finite(),
        "carrier cross space must be finite"
    );
    assert!(
        context.flank_depth.is_finite(),
        "carrier flank depth must be finite"
    );
    assert!(
        context.goal_progress.is_finite(),
        "carrier goal progress must be finite"
    );
    assert!(
        context.dribble_space.is_finite(),
        "carrier dribble space must be finite"
    );

    let distance_falloff = context.goal_distance / context.shoot_range;
    let mut options: Vec<brain::CarrierOption> = vec![
        brain::CarrierOption {
            id: "shoot".to_string(),
            kind: "shoot".to_string(),
            score: 100.0 - distance_falloff * distance_falloff * 60.0
                + clamp_unit(context.angle_quality) * 25.0
                - clamp_unit(context.keeper_coverage) * 12.0
                + clamp_unit(context.space) * 18.0,
            payload: None,
            reference: None,
        },
        brain::CarrierOption {
            id: "dribble".to_string(),
            kind: "dribble".to_string(),
            score: 10.0
                + clamp_unit(context.dribble_space) * 35.0
                + clamp_unit(context.goal_progress) * 25.0,
            payload: None,
            reference: None,
        },
    ];

    if let Some(cross_target) = context.cross_target
        && context.box_targets > 0
    {
        options.push(brain::CarrierOption {
            id: "cross".to_string(),
            kind: "cross".to_string(),
            score: 60.0
                + clamp_unit(context.flank_depth) * 45.0
                + f64::from(context.box_targets.min(3)) * 8.0
                + clamp_unit(context.cross_space) * 15.0,
            payload: None,
            reference: Some(brain::OptionReference::Index(cross_target)),
        });
    }

    let mut seen: Vec<u32> = Vec::new();
    for (index, pass) in context.passes.iter().enumerate() {
        assert!(
            pass.player_index >= 1,
            "pass option {} player must be a positive integer",
            index + 1
        );
        assert!(
            !seen.contains(&pass.player_index),
            "pass option players must be unique"
        );
        seen.push(pass.player_index);
        assert!(
            pass.openness.is_finite(),
            "pass option openness must be finite"
        );
        assert!(
            pass.forward_progress.is_finite(),
            "pass option forward progress must be finite"
        );
        assert!(
            pass.distance.is_finite(),
            "pass option distance must be finite"
        );
        assert!(
            pass.distance >= 0.0,
            "pass option distance must be non-negative"
        );
        if let Some(lane_fraction) = pass.lane_fraction {
            assert!(
                lane_fraction.is_finite(),
                "pass option lane fraction must be finite"
            );
        }
        let payload = pass.lane_fraction.map(|lane_fraction| {
            let mut map = IndexMap::new();
            map.insert(
                "lane_fraction".to_string(),
                brain::BrainPayloadValue::Number(lane_fraction),
            );
            map
        });
        options.push(brain::CarrierOption {
            id: format!("pass_{}", pass.player_index),
            kind: "pass".to_string(),
            score: 18.0
                + (pass.openness.max(0.0) / 150.0).min(1.0) * 45.0
                + (pass.forward_progress / 300.0).clamp(-1.0, 1.0) * 28.0
                - (pass.distance / 420.0).min(1.0) * 18.0
                + (PASS_SPACE_BALANCE - clamp_unit(context.space)) * PASS_SPACE_WEIGHT
                - if pass.interception_risk {
                    PASS_RISK_PENALTY
                } else {
                    0.0
                },
            payload,
            reference: Some(brain::OptionReference::Index(pass.player_index)),
        });
    }
    options
}

/// Decide what the carrier does: at the authored high-composure boundary
/// the temperature is exactly zero, so an effectively settled choice takes
/// the argmax without needlessly advancing its dedicated decision RNG
/// stream.
#[must_use]
pub fn decide_carrier(
    options: &[brain::CarrierOption],
    composure: f64,
    pressure: f64,
    rng_state: u32,
) -> (brain::CarrierOption, u32) {
    let base_temperature = if composure >= SHARP_COMPOSURE {
        0.0
    } else {
        BASE_TEMPERATURE
    };
    brain::decide_carrier(options, composure, pressure, base_temperature, rng_state)
}
